### PR TITLE
Redis and custom counter support

### PR DESCRIPTION
**Pull request by jeremy**
Built-in support for a Redis backend. Allow providing your own counter, such as an in-process expiring hash → hash → count.

### CODE REVIEW

**Comment by qrush** [comment]
Heck to the yes. :+1:

File: lib/rack/ratelimit.rb [hunk 1]
```diff
@@ -163,5 +175,23 @@ def increment(classification, timestamp)
...
+
+      # Increment the request counter and return the current count.
+      def increment(classification, timestamp)
+        key = 'rack-ratelimit/%s/%s/%i' % [@name, classification, timestamp]
+
+        # Returns [count, expire_ok] response for each multi command.
+        # Return the first, the count.
+        @redis.multi do |redis|
+          redis.incr key
+          redis.expire key, @period
```
Comment (qrush) [inline lib/rack/ratelimit.rb]: This will reset the TTL every time. Is that intentional?

Comment (qrush) [inline lib/rack/ratelimit.rb]: redis-cli:

```
127.0.0.1:6379> INCR bar
(integer) 1
127.0.0.1:6379> EXPIRE bar 60
(integer) 1
127.0.0.1:6379> TTL bar
(integer) 57
127.0.0.1:6379> INCR bar
(integer) 2
127.0.0.1:6379> EXPIRE bar 60
(integer) 1
127.0.0.1:6379> TTL bar
(integer) 58
```

Comment (jeremy) [inline lib/rack/ratelimit.rb]: Correct. By setting the expiry unconditionally, we don't have to care about the current key state. Once we move to the next timeslice, the key will change and the old key will naturally expire.

Comment (qrush) [inline lib/rack/ratelimit.rb]: Ah ha- that's the piece I'm missing. We aren't relying on the cache's knowledge of time, the library handles that. The expire time is to clean up keys as we go along. :recycle:

**Pull request merged**